python/pylibcugraph/pylibcugraph/biased_neighbor_sample.pyx

```diff
@@ -0,0 +1,448 @@
+    cugraph_prior_sources_behavior_t,
+    cugraph_compression_type_t,
+    cugraph_sampling_options_t,
+    cugraph_sampling_options_create,
+    cugraph_sampling_options_free,
+    cugraph_sampling_set_with_replacement,
+    cugraph_sampling_set_return_hops,
+    cugraph_sampling_set_prior_sources_behavior,
+    cugraph_sampling_set_dedupe_sources,
+    cugraph_sampling_set_renumber_results,
+    cugraph_sampling_set_compress_per_hop,
+    cugraph_sampling_set_compression_type,
+    cugraph_sampling_set_retain_seeds,
+)
+from pylibcugraph._cugraph_c.sampling_algorithms cimport (
+    cugraph_biased_neighbor_sample,
+
+)
+from pylibcugraph.resource_handle cimport (
+    ResourceHandle,
+)
+from pylibcugraph.graphs cimport (
+    _GPUGraph,
+)
+from pylibcugraph.utils cimport (
+    assert_success,
+    assert_CAI_type,
+    assert_AI_type,
+    get_c_type_from_numpy_type,
+)
+from pylibcugraph.internal_types.sampling_result cimport (
+    SamplingResult,
+)
+from pylibcugraph._cugraph_c.random cimport (
+    cugraph_rng_state_t
+)
+from pylibcugraph.random cimport (
+    CuGraphRandomState
+)
+import warnings
+
+# TODO accept cupy/numpy random state in addition to raw seed.
+def biased_neighbor_sample(ResourceHandle resource_handle,
+                            _GPUGraph input_graph,
+                            start_list,
+                            h_fan_out,
+                            *,
+                            bool_t with_replacement,
+                            bool_t do_expensive_check,
+                            with_edge_properties=True,
+                            batch_id_list=None,
+                            label_list=None,
+                            label_to_output_comm_rank=None,
+                            label_offsets=None,
+                            biases=None,
+                            prior_sources_behavior=None,
+                            deduplicate_sources=False,
+                            return_hops=False,
+                            renumber=False,
+                            retain_seeds=False,
+                            compression='COO',
+                            compress_per_hop=False,
+                            random_state=None,
+                            return_dict=False,):
+    """
+    Does neighborhood sampling, which samples nodes from a graph based on the
+    current node's neighbors, with a corresponding fanout value at each hop.
+
+    Parameters
+    ----------
+    resource_handle: ResourceHandle
+        Handle to the underlying device and host resources needed for
+        referencing data and running algorithms.
+
+    input_graph : SGGraph or MGGraph
+        The input graph, for either Single or Multi-GPU operations.
+
+    start_list: device array type
+        Device array containing the list of starting vertices for sampling.
+
+    h_fan_out: numpy array type
+        Device array containing the brancing out (fan-out) degrees per
```

Comment:
Is this a device array?  It has an `h_` prefix and the assertion below looks for (non-CUDA) array_interface.

Reply:
This is a host array.  I naively copied this from `uniform_neighbor_sample` so I'll fix the docstring there too.